runtime/src/bank.rs

```diff
@@ -691,77 +691,106 @@ impl Bank {
+
+        // this sysvar could be retired...
         self.update_sysvar_account(&sysvar::rewards::id(), |account| {
             sysvar::rewards::create_account(
                 self.inherit_sysvar_account_balance(account),
                 validator_point_value,
             )
         });
 
-        let validator_rewards = self.pay_validator_rewards(validator_point_value);
+        let validator_rewards_paid =
+            self.stakes.read().unwrap().vote_balance_and_staked() - vote_balance_and_staked;
+
+        // verify that we didn't pay any more than we expected to
+        assert!(validator_rewards >= validator_rewards_paid);
 
         self.capitalization
-            .fetch_add(validator_rewards as u64, Ordering::Relaxed);
+            .fetch_add(validator_rewards_paid, Ordering::Relaxed);
     }
 
     /// iterate over all stakes, redeem vote credits for each stake we can
     ///   successfully load and parse, return total payout
-    fn pay_validator_rewards(&mut self, point_value: f64) -> u64 {
+    fn pay_validator_rewards(&mut self, rewards: u64) -> f64 {
         let stake_history = self.stakes.read().unwrap().history().clone();
-        let mut validator_rewards = HashMap::new();
 
-        let total_validator_rewards = self
-            .stake_delegations()
+        let stake_delegations = self.stake_delegations();
+
+        // first, count total points
+        let points: u128 = stake_delegations
             .iter()
             .map(|(stake_pubkey, delegation)| {
+                match (
+                    self.get_account(&stake_pubkey),
+                    self.get_account(&delegation.voter_pubkey),
+                ) {
+                    (Some(stake_account), Some(vote_account)) => stake_state::calculate_points(
+                        &stake_account,
+                        &vote_account,
+                        Some(&stake_history),
+                    )
+                    .unwrap_or(0),
+                    (_, _) => 0,
+                }
+            })
+            .sum();
+
+        if points == 0 {
+            return 0.0;
+        }
+
+        let point_value = PointValue { rewards, points };
+
+        let mut rewards = HashMap::new();
+
+        // pay according to point value
+        stake_delegations
+            .iter()
+            .for_each(|(stake_pubkey, delegation)| {
                 match (
                     self.get_account(&stake_pubkey),
                     self.get_account(&delegation.voter_pubkey),
                 ) {
                     (Some(mut stake_account), Some(mut vote_account)) => {
-                        let rewards = stake_state::redeem_rewards(
+                        let redeemed = stake_state::redeem_rewards(
                             &mut stake_account,
                             &mut vote_account,
-                            point_value,
+                            &point_value,
                             Some(&stake_history),
```

Comment:
nits: Not related to this pr's change per se, but I'd like to change this to take `&StakeHistory` with similar spirit of [this](https://github.com/solana-labs/solana/pull/10914/files#r451370146).